extensions/register.md

```diff
@@ -0,0 +1,147 @@
+to follow the precedent set by [the multiline draft][multiline].
+
+The defined keys are:
+
+ * `before-connect` - if present, indicates the server supports early registration, so it
+   MUST NOT use `FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`
+ * `email-required` - if present, registrations require a valid email address to process
+
+Clients MUST ignore any value assigned to these keys, and MUST ignore any unknown key.
+
+### Commands
+
+    REGISTER {<email> | "*"} <password>
+    
+The `REGISTER` command informs the server of a request to register an account named for the current
+nick of the requestor.
+
+The `REGISTER` command MAY be sent at any point during the connection that the client has a valid
+nickname. If the client sends `REGISTER` before completing connection registration, and receives a
+`FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`, it SHOULD make a second attempt after it receives the
+welcome message.
+
+    VERIFY <account> <code>
+    
+The `VERIFY` command completes a registration that required email verification.
+
+### Responses
+
+    REGISTER SUCCESS <account> <message>
+    VERIFY SUCCESS <account> <message>
+    
+Sent by the server when the `REGISTER` or `VERIFY` fully succeeded, and no
+further action is needed.
+Afterward, the client is assumed to be authenticated to the `<account>`
+as if it used SASL.
+
+    REGISTER VERIFICATION_REQUIRED <account> <message>
+    
+Sent by the server as a response to `REGISTER` when further action is required
+to complete registration of the `<account>`, as explain in the human-readable `<message>`.
+
+    FAIL REGISTER USERNAME_EXISTS <account> <message>
+
+Sent by the server as a response to `REGISTER` when the client tried to register
+while using a nick that is not available as a new account's name
+(for example, because someone already registered it).
+
+The server MUST NOT send this response before the client sent a `NICK` command.
+
+The client MAY try registering again later.
+Clients should not automatically retry immediately without changing their nickname
+or waiting.
+
+    FAIL REGISTER ALREADY_REGISTERED <account> <message>
+    FAIL VERIFY ALREADY_REGISTERED <account> <message>
+
+Sent by the server is the client is already authenticated.
+
+    FAIL REGISTER WEAK_PASSWORD <account> <message>
+
+Sent by the server if the password is considered too weak.
+
+The client MAY try registering again.
+
+    FAIL REGISTER UNACCEPTABLE_PASSWORD <account> <message>
+
+Sent by the server if the password is invalid for any reason other than weakness.
+
+The client MAY try registering again.
+
+    FAIL REGISTER INVALID_EMAIL <account> <message>
+
+Sent by the server if it cannot send emails to the provided address.
+
+    FAIL REGISTER UNACCEPTABLE_EMAIL <account> <message>
+
+Sent by the server if the email address is valid, but not available for account registration.
+
+    FAIL REGISTER COMPLETE_CONNECTION_REQUIRED
+    FAIL VERIFY COMPLETE_CONNECTION_REQUIRED
```

Comment:
You'll need the `<message>` param on the end of these to match the std replies format.